common/src/app/common/files/validate.cljc

```diff
@@ -178,9 +184,21 @@
                          (str/ffmt "Main instance id of component % is not valid" (:component-id shape))
                          shape file page))
         (when-not (= (:main-instance-page component) (:id page))
-          (report-error! :invalid-main-instance-page
-                         (str/ffmt "Main instance page of component % is not valid" (:component-id shape))
-                         shape file page))))))
+          (let [component-page (ctf/get-component-page (:data file) component)
+                main-component (ctst/get-shape component-page (:main-instance-id component))]
+            ;; We must check if the same component has main instances in different pages. 
+            ;; In that case one of those instances shouldn't be main
+            (if (and
+                  (not= (:id page) (:id component-page))
+                  (= (:main-instance main-component) (:main-instance shape))
+                  (= (:component-id main-component) (:component-id shape)))
```

Comment:
If we have come to here, we already know that the component page is different, and the instance-id is irrelevant. I think it will be enough to check if `main-component` exists. If so, we know that the current shape is not expected to be a main.